### PR TITLE
add support for local metadata (option 1)

### PR DESCRIPTION
**Pull request by tomerd**
motivation: allow users to define metadata at the log API call site

changes: add metadata param to all log APIs, and implement merge at our example implementations

### CODE REVIEW

**Comment by tomerd** [comment]
@weissi not sure about this one, since the merging is delegated to the implementation, but given the current api design where metadata is managed at the logger level, this seems to be the most flexible design

the alternative is to do the merging at the API level and then the implementations will need to know they should ignore the instance state and only use the merged one that came through the LogHandler.log call. see [option 2](https://github.com/weissi/swift-server-logging-api-proposal/pull/17). the main advantage of option 2 is somewhat easier API to reason about, the disadvantage is performance hit on prettyfing the metadata every call 

wdyt?

/cc @ktoso

File: Tests/LoggingTests/TestLogger.swift [hunk 1]
```diff
@@ -30,15 +30,11 @@ internal struct TestLogger: LogHandler {
...
+    func log(level: Logging.Level, message: String, metadata: Logging.Metadata?, error: Error?, file: StaticString, function: StaticString, line: UInt) {
+        let metadata = (self._metadataSet ? self.metadata : MDC.global.metadata).merging(metadata ?? [:], uniquingKeysWith: { _, new in new })
         var l = logger // local copy since we gonna override its metadata
         l.metadata = metadata
-        if let e = error {
-            l.log(level: level, message: message, error: e, file: file, function: function, line: line)
-        } else {
-            l.log(level: level, message: message, file: file, function: function, line: line)
-        }
+        l.log(level: level, message: message, metadata: metadata, error: error, file: file, function: function, line: line)
```
Comment (tomerd) [inline Tests/LoggingTests/TestLogger.swift:37]: unrelated cruft that i removed

File: Sources/Logging/Logging.swift [hunk 2]
```diff
@@ -10,7 +10,7 @@ public protocol LogHandler {
     // `Logger`'s `info`, `error`, or `warning` functions.
     //
     // An implementation does not need to check the log level because that has been done before by `Logger` itself.
-    func log(level: Logging.Level, message: String, error: Error?, file: StaticString, function: StaticString, line: UInt)
+    func log(level: Logging.Level, message: String, metadata: Logging.Metadata?, error: Error?, file: StaticString, function: StaticString, line: UInt)
```
[the remark this answers is not in the document]
Comment (tomerd) [reply Sources/Logging/Logging.swift]: see options 2: https://github.com/weissi/swift-server-logging-api-proposal/pull/17

File: Sources/ExampleImplementation/SimpleLogger.swift [hunk 3]
```diff
@@ -21,8 +21,9 @@ internal final class SimpleLogger {
         self.formatter = formatter
     }
 
-    public func log(level: Logging.Level, message: String, error: Error?, printer: (String) -> Void) {
-        printer("[\(self.label)] \(self.formatter.string(from: Date()))\(self.prettyMetadata.map { " \($0)" } ?? "") \(level): \(message)\(error.map { " \($0)" } ?? "")")
+    public func log(level: Logging.Level, message: String, metadata: Logging.Metadata?, error: Error?, printer: (String) -> Void) {
+        let prettyMetadata = metadata?.isEmpty ?? true ? self.prettyMetadata : self.prettify(self.metadata.merging(metadata!, uniquingKeysWith: { _, new in new }))
```
Comment (tomerd) [inline Sources/ExampleImplementation/SimpleLogger.swift:25]: with this option, the implementation does the merging as demonstrated here

**Comment by ktoso** [comment]
Looked in depth at both options and I think `option 1` (this one) is viable and good™.

It makes sense to leave to implementations to do what they want with the incoming per log statement metadata. The ability to keep the pre-rendered if-no-per-log-statement-metadata is also very valuable (or rather, not being able to do so in option 2 disqualifies it I'd say even).

👍 for this version. And guidance that implementations should optimize for the metadata staying stable, and merge on incoming more metadata, though they MAY choose to do other things.

**Comment by weissi** [comment]
thanks @tomerd , that lgtm!

**Pull request merged**